### ISSUE TITLE
False positive for `Lint/RedundantSafeNavigation` with `rescue`

### ISSUE THREAD

**Issue by zverok** [rubocop/rubocop#15148]
### Before submitting

- [x] I have searched the existing issues and this hasn't been reported yet.
- [x] I am running the latest stable version of RuboCop.

### Issue type

False positive

### Cop name

Lint/RedundantSafeNavigation

### Steps to reproduce the problem

example.rb:
```ruby
x = nil
begin
  x.length
rescue
  x&.first
end
```

### Expected behavior

No linting problems

### Actual behavior

```
example.rb:5:4: W: [Correctable] Lint/RedundantSafeNavigation: Redundant safe navigation on non-nil receiver (detected by analyzing previous code/method invocations).
  x&.first
   ^^
```
This is wrong, as the exception might've been `NoMethodError` exactly due to `x` being `nil`

### RuboCop configuration

```yaml
Lint/RedundantSafeNavigation:
  InferNonNilReceiver: true
```

### RuboCop version

```shell
1.86.1 (using Parser 3.3.11.1, rubocop-ast 1.49.1, analyzing as Ruby 3.1, running on ruby 3.3.7) [x86_64-linux]

(Actually running from the latest `master`)
```
